File: leetcode js/326. Power of Three.js
// 题目说明：
// Given an integer, write a function to determine if it is a power of three.
// 用递归和log两种方法


// 方法一：
/**
 * @param {number} n
 * @return {boolean}
 */
 var isPowerOfThree = function(n) {
     console.log(n);
     if(n === 1) {
         console.log("a");
         return true;
     } else if(n === 0){
         return false;
     } else if( n % 3 === 0){
        //  注意做的时候这里没加return，导致返回的是undefined
         return isPowerOfThree (n / 3);
     } else {
         return false;
     }
 };

 var x = isPowerOfThree(27);
 console.log(x);
// 方法一：
// 3^x=n
// log(3^x) = log(n)
// x log(3) = log(n)
// x = log(n) / log(3)

 var isPowerOfThree = function(n) {
    var temp = Math.log(n)/Math.log(3);
    return Math.abs(temp - Math.round(temp))< 0.0000000001;

};

File: leetcode java/Math/326. Power of Three.java
/*
Given an integer, write a function to determine if it is a power of three.
1.递归
2.数学方法
x = log(n)/log(3),x是整数就可以了
--那么判断整数方法：double类型，则绝对值差小于10e-15；floor类型，则绝对值差小于10e-6
 */
 public class Solution {
     public boolean isPowerOfThree(int n) {
         if(n == 1) {
             return true;
         } else if(n == 0) {
             return false;
         } else if(n % 3 == 0) {
             return isPowerOfThree(n/3);
         } else {
             return false;
         }
     }
 }

 public class Solution {
     public boolean isPowerOfThree(int n) {
         if(n == 0) {
             return false;
         }
         double x = Math.log(n)/Math.log(3);

         return Math.abs(x - Math.round(x)) < 10e-15;
     }
 }
